## Wrangle "@weratedogs" twitter data
The objective of this document is to wrangle the tweet archive of @weratedogs user on twitter. To inspect the data and augment it with other sources of information including by fetching data from twitter API. Once the data is gathered from all sources, the data will be inspected and cleaned.

In [99]:
# import necessary libraries
import pandas as pd
import numpy as np
import requests # to make http requests
import tweepy # to work with the twitter api

import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

In [65]:
# file and URL names to process
twitter_archive_file = 'twitter-archive-enhanced.csv'
image_predictions_url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
tweets_file = 'tweet_json.txt'

### Gather
Get data from the initially known data sources

In [5]:
# get data from WeRateDogs twitter archive
twitter_archive = pd.read_csv(twitter_archive_file)
twitter_archive.head(1)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None

In [12]:
# get tweet image predictions data from url
image_predictions = pd.read_csv(image_predictions_url, sep='\t')
image_predictions.head(1)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   

   img_num                      p1   p1_conf  p1_dog      p2   p2_conf  \
0        1  Welsh_springer_spaniel  0.465074    True  collie  0.156665   

   p2_dog                 p3   p3_conf  p3_dog  
0    True  Shetland_sheepdog  0.061428    True

### Assess
Understand the obtained data a little before fetching the tweets

In [13]:
twitter_archive.shape

(2356, 17)

In [15]:
image_predictions.shape

(2075, 12)

### Gather
__`twitter_archive` table__ has more lines of data, hence let us fetch the tweets based on this

In [29]:
# get keys and secrets from the environment
import os
consumer_key = os.environ['TWITTER_API_KEY']
consumer_secret = os.environ['TWITTER_API_SECRET']
access_token = os.environ['TWITTER_ACCESS_TOKEN']
access_token_secret = os.environ['TWITTER_ACCESS_SECRET']

In [86]:
# Initialize the tweepy library 
import tweepy
import json
auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_token_secret)
api = tweepy.API(auth)

In [76]:
# get the full tweet json for each tweet and write to a file one line per tweet
with open(tweets_file, 'w') as file:
    # statuses_lookup can fetch 100 tweets at a time
    # take 100 tweets at a time from twitter_archive
    for index in range(0, len(twitter_archive), 100):
        id_list = twitter_archive.tweet_id.iloc[index:index+100]
        try:
            statuses = api.statuses_lookup(id_list.array)
            for status in statuses:
                file.write(json.dumps(status._json) + '\n')
        except Exception as e:
            print(str(e))

In [85]:
# read this data into a dataframe
detailed_tweets = pd.read_json(tweets_file, lines=True)
detailed_tweets.head(1)

contributors  coordinates          created_at  \
0           NaN          NaN 2017-06-18 16:57:37   

                                            entities extended_entities  \
0  {'hashtags': [], 'symbols': [], 'user_mentions...               NaN   

   favorite_count  favorited  geo                  id              id_str  \
0           18150      False  NaN  876484053909872640  876484053909872640   

   ... quoted_status  quoted_status_id  quoted_status_id_str  retweet_count  \
0  ...           NaN               NaN                   NaN           2268   

   retweeted  retweeted_status  \
0      False               NaN   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   

                                                text  truncated  \
0  This is Benedict. He wants to thank you for th...       True   

                                                user  
0  {'id': 4196983835, 'id_str': '4196983835', 'na...  

[1 rows x 29 columns]

### Assess

#### Analyze the `twitter_archive` data

In [77]:
twitter_archive.sample(10)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
496   813157409116065792                    NaN                  NaN   
720   783347506784731136                    NaN                  NaN   
1109  733828123016450049                    NaN                  NaN   
916   756998049151549440                    NaN                  NaN   
1332  705475953783398401                    NaN                  NaN   
1596  686286779679375361                    NaN                  NaN   
1623  684902183876321280                    NaN                  NaN   
725   782722598790725632                    NaN                  NaN   
1557  688804835492233216                    NaN                  NaN   
2165  669367896104181761                    NaN                  NaN   

                      timestamp  \
496   2016-12-25 23:00:08 +0000   
720   2016-10-04 16:46:14 +0000   
1109  2016-05-21 01:13:53 +0000   
916   2016-07-23 23:42:53 +0000   
1332  2016-03-03 19:32:29 +0000   
1596  2016-01-10 20:41:33 +0000   
1623  2016-01-07 00:59:40 +0000   
725   2016-10-02 23:23:04 +0000   
1557  2016-01-17 19:27:24 +0000   
2165  2015-11-25 04:11:57 +0000   

                                                 source  \
496   <a href="http://twitter.com/download/iphone" r...   
720   <a href="http://twitter.com/download/iphone" r...   
1109  <a href="http://twitter.com/download/iphone" r...   
916   <a href="http://twitter.com/download/iphone" r...   
1332  <a href="http://twitter.com/download/iphone" r...   
1596  <a href="http://vine.co" rel="nofollow">Vine -...   
1623  <a href="http://twitter.com/download/iphone" r...   
725   <a href="http://twitter.com/download/iphone" r...   
1557  <a href="http://twitter.com/download/iphone" r...   
2165  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
496   This is Layla. It is her first Christmas. She ...                  NaN   
720   RT @dog_rates: This is Kenny. He just wants to...         6.742918e+17   
1109  This is Terry. The harder you hug him the fart...                  NaN   
916   This is Oliver. He's an English Creamschnitzel...                  NaN   
1332  Say hello to Zara. She found a sandal and coul...                  NaN   
1596  When bae calls your name from across the room....                  NaN   
1623  This is Perry. He's an Augustus Gloopster. Ver...                  NaN   
725   This is Penny. She fought a bee and the bee wo...                  NaN   
1557  When you stumble but recover quickly cause you...                  NaN   
2165  This is Chip. Chip's pretending to be choked. ...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
496                        NaN                        NaN   
720               4.196984e+09  2015-12-08 18:17:56 +0000   
1109                       NaN                        NaN   
916                        NaN                        NaN   
1332                       NaN                        NaN   
1596                       NaN                        NaN   
1623                       NaN                        NaN   
725                        NaN                        NaN   
1557                       NaN                        NaN   
2165                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
496   https://twitter.com/dog_rates/status/813157409...                12   
720   https://twitter.com/dog_rates/status/674291837...                11   
1109  https://twitter.com/dog_rates/status/733828123...                10   
916   https://twitter.com/dog_rates/status/756998049...                11   
1332  https://twitter.com/dog_rates/status/705475953...                12   
1596                      https://vine.co/v/iMZx6aDbExn                12   
1623  https://twitter.com/dog_rates/status/684902183...                11   
725   https://twit

In [88]:
twitter_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [90]:
twitter_archive.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [126]:
twitter_archive[twitter_archive.rating_numerator < 10].sample(10)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
2255  667773195014021121                    NaN                  NaN   
1861  675483430902214656                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
1735  679729593985699840                    NaN                  NaN   
1756  678767140346941444                    NaN                  NaN   
1484  693231807727280129                    NaN                  NaN   
1978  672984142909456390                    NaN                  NaN   
2302  667012601033924608                    NaN                  NaN   
1803  676948236477857792                    NaN                  NaN   
2242  667911425562669056                    NaN                  NaN   

                      timestamp  \
2255  2015-11-20 18:35:10 +0000   
1861  2015-12-12 01:12:54 +0000   
2347  2015-11-16 00:55:59 +0000   
1735  2015-12-23 18:25:38 +0000   
1756  2015-12-21 02:41:11 +0000   
1484  2016-01-30 00:38:37 +0000   
1978  2015-12-05 03:41:37 +0000   
2302  2015-11-18 16:12:51 +0000   
1803  2015-12-16 02:13:31 +0000   
2242  2015-11-21 03:44:27 +0000   

                                                 source  \
2255  <a href="http://twitter.com" rel="nofollow">Tw...   
1861  <a href="http://twitter.com/download/iphone" r...   
2347  <a href="http://twitter.com/download/iphone" r...   
1735  <a href="http://twitter.com/download/iphone" r...   
1756  <a href="http://twitter.com/download/iphone" r...   
1484  <a href="http://twitter.com/download/iphone" r...   
1978  <a href="http://twitter.com/download/iphone" r...   
2302  <a href="http://twitter.com/download/iphone" r...   
1803  <a href="http://twitter.com/download/iphone" r...   
2242  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
2255  This is a rare Hungarian Pinot named Jessiga. ...                  NaN   
1861  Rare shielded battle dog here. Very happy abou...                  NaN   
2347  My oh my. This is a rare blond Canadian terrie...                  NaN   
1735  This is Hunter. He was playing with his ball m...                  NaN   
1756  This is Mia. She makes awful decisions. 8/10 h...                  NaN   
1484  This is Bodie. He's not proud of what he did, ...                  NaN   
1978  Very happy pup here. Always smiling. Loves his...                  NaN   
2302  This is Klevin. He laughs a lot. Very cool dog...                  NaN   
1803  This is Karl. Karl thinks he's slick. 6/10 sne...                  NaN   
2242  Wow. Armored dog here. Ready for battle. Face ...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
2255                       NaN                        NaN   
1861                       NaN                        NaN   
2347                       NaN                        NaN   
1735                       NaN                        NaN   
1756                       NaN                        NaN   
1484                       NaN                        NaN   
1978                       NaN                        NaN   
2302                       NaN                        NaN   
1803                       NaN                        NaN   
2242                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
2255  https://twitter.com/dog_rates/status/667773195...                 8   
1861  https://twitter.com/dog_rates/status/675483430...                 5   
2347  https://twitter.com/dog_rates/status/666057090...                 9   
1735  https://twitter.com/dog_rates/status/679729593...                 8   
1756  https://twitter.com/dog_rates/status/678767140...                 8   
1484  https://twitter.com/dog_rates/status/693231807...                 9   
1978  https://twitter.com/dog_rates/status/672984142...                 9   
2302  https://twit

In [112]:
twitter_archive.iloc[315].expanded_urls

'https://twitter.com/dog_rates/status/835152434251116546/photo/1,https://twitter.com/dog_rates/status/835152434251116546/photo/1,https://twitter.com/dog_rates/status/835152434251116546/photo/1'

In [120]:
twitter_archive.iloc[2317].expanded_urls

'https://twitter.com/dog_rates/status/666644823164719104/photo/1'

In [128]:
twitter_archive[twitter_archive.rating_numerator < 10].shape

(440, 17)

In [113]:
twitter_archive[twitter_archive.rating_numerator > 15].sample(10)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1120  731156023742988288                    NaN                  NaN   
1712  680494726643068929                    NaN                  NaN   
1351  704054845121142784                    NaN                  NaN   
313   835246439529840640           8.352460e+17           26259576.0   
2074  670842764863651840                    NaN                  NaN   
433   820690176645140481                    NaN                  NaN   
188   855862651834028034           8.558616e+17          194351775.0   
1254  710658690886586372                    NaN                  NaN   
1779  677716515794329600                    NaN                  NaN   
1228  713900603437621249                    NaN                  NaN   

                      timestamp  \
1120  2016-05-13 16:15:54 +0000   
1712  2015-12-25 21:06:00 +0000   
1351  2016-02-28 21:25:30 +0000   
313   2017-02-24 21:54:03 +0000   
2074  2015-11-29 05:52:33 +0000   
433   2017-01-15 17:52:40 +0000   
188   2017-04-22 19:15:32 +0000   
1254  2016-03-18 02:46:49 +0000   
1779  2015-12-18 05:06:23 +0000   
1228  2016-03-27 01:29:02 +0000   

                                                 source  \
1120  <a href="http://twitter.com/download/iphone" r...   
1712  <a href="http://twitter.com/download/iphone" r...   
1351  <a href="http://twitter.com/download/iphone" r...   
313   <a href="http://twitter.com/download/iphone" r...   
2074  <a href="http://twitter.com/download/iphone" r...   
433   <a href="http://twitter.com/download/iphone" r...   
188   <a href="http://twitter.com/download/iphone" r...   
1254  <a href="http://twitter.com/download/iphone" r...   
1779  <a href="http://twitter.com/download/iphone" r...   
1228  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
1120  Say hello to this unbelievably well behaved sq...                  NaN   
1712  Here we have uncovered an entire battalion of ...                  NaN   
1351  Here is a whole flock of puppers.  60/50 I'll ...                  NaN   
313   @jonnysun @Lin_Manuel ok jomny I know you're e...                  NaN   
2074  After so many requests... here you go.\n\nGood...                  NaN   
433   The floofs have been released I repeat the flo...                  NaN   
188   @dhmontgomery We also gave snoop dogg a 420/10...                  NaN   
1254  Here's a brigade of puppers. All look very pre...                  NaN   
1779  IT'S PUPPERGEDDON. Total of 144/120 ...I think...                  NaN   
1228  Happy Saturday here's 9 puppers on a bench. 99...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
1120                       NaN                        NaN   
1712                       NaN                        NaN   
1351                       NaN                        NaN   
313                        NaN                        NaN   
2074                       NaN                        NaN   
433                        NaN                        NaN   
188                        NaN                        NaN   
1254                       NaN                        NaN   
1779                       NaN                        NaN   
1228                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
1120  https://twitter.com/dog_rates/status/731156023...               204   
1712  https://twitter.com/dog_rates/status/680494726...                26   
1351  https://twitter.com/dog_rates/status/704054845...                60   
313                                                 NaN               960   
2074  https://twitter.com/dog_rates/status/670842764...               420   
433   https://twitter.com/dog_rates/status/820690176...                84   
188                                                 NaN               420   
1254  https://twit

In [110]:
twitter_archive.iloc[979].expanded_urls

'https://twitter.com/dog_rates/status/749981277374128128/photo/1'

In [129]:
twitter_archive[twitter_archive.rating_numerator > 15].shape

(26, 17)

In [130]:
twitter_archive[twitter_archive.name == "None"].sample(10)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1459  695064344191721472                    NaN                  NaN   
1111  733482008106668032                    NaN                  NaN   
195   855138241867124737                    NaN                  NaN   
1689  681340665377193984           6.813394e+17         4.196984e+09   
1654  683449695444799489                    NaN                  NaN   
1224  714214115368108032                    NaN                  NaN   
1506  691756958957883396                    NaN                  NaN   
1068  740373189193256964                    NaN                  NaN   
1263  710117014656950272                    NaN                  NaN   
411   823269594223824897                    NaN                  NaN   

                      timestamp  \
1459  2016-02-04 02:00:27 +0000   
1111  2016-05-20 02:18:32 +0000   
195   2017-04-20 19:16:59 +0000   
1689  2015-12-28 05:07:27 +0000   
1654  2016-01-03 00:47:59 +0000   
1224  2016-03-27 22:14:49 +0000   
1506  2016-01-25 22:58:05 +0000   
1068  2016-06-08 02:41:38 +0000   
1263  2016-03-16 14:54:24 +0000   
411   2017-01-22 20:42:21 +0000   

                                                 source  \
1459  <a href="http://twitter.com/download/iphone" r...   
1111  <a href="http://twitter.com/download/iphone" r...   
195   <a href="http://twitter.com/download/iphone" r...   
1689  <a href="http://twitter.com/download/iphone" r...   
1654  <a href="http://twitter.com/download/iphone" r...   
1224  <a href="http://twitter.com/download/iphone" r...   
1506  <a href="http://twitter.com/download/iphone" r...   
1068  <a href="http://twitter.com/download/iphone" r...   
1263  <a href="http://twitter.com/download/iphone" r...   
411   <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
1459  This may be the greatest video I've ever been ...                  NaN   
1111  "Ello this is dog how may I assist" ...10/10 h...                  NaN   
195   RT @frasercampbell_: oh my... what's that... b...         8.551225e+17   
1689  I've been told there's a slight possibility he...                  NaN   
1654  I just want to be friends with this dog. Appea...                  NaN   
1224  Happy Easter from the squad! 🐇🐶 13/10 for all ...                  NaN   
1506  THE BRITISH ARE COMING\nTHE BRITISH ARE COMING...                  NaN   
1068  After so many requests, this is Bretagne. She ...                  NaN   
1263  This pupper got her hair chalked for her birth...                  NaN   
411   RT @dog_rates: We only rate dogs. Please don't...         8.222448e+17   

      retweeted_status_user_id retweeted_status_timestamp  \
1459                       NaN                        NaN   
1111                       NaN                        NaN   
195               7.475543e+17  2017-04-20 18:14:33 +0000   
1689                       NaN                        NaN   
1654                       NaN                        NaN   
1224                       NaN                        NaN   
1506                       NaN                        NaN   
1068                       NaN                        NaN   
1263                       NaN                        NaN   
411               4.196984e+09  2017-01-20 00:50:15 +0000   

                                          expanded_urls  rating_numerator  \
1459  https://twitter.com/dog_rates/status/695064344...                 4   
1111  https://twitter.com/dog_rates/status/733482008...                10   
195   https://twitter.com/frasercampbell_/status/855...                14   
1689                                                NaN                 5   
1654  https://twitter.com/dog_rates/status/683449695...                10   
1224  https://twitter.com/dog_rates/status/714214115...                13   
1506  https://twitter.com/dog_rates/status/691756958...                10   
1068  https://twit

In [124]:
twitter_archive[twitter_archive.name == "None"].shape

(745, 17)

#### Analyze the `image_predictions` data

In [122]:
image_predictions.sample(10)

tweet_id                                          jpg_url  \
2064  890006608113172480  https://pbs.twimg.com/media/DFnwSY4WAAAMliS.jpg   
460   674793399141146624  https://pbs.twimg.com/media/CV1ZA3oWEAA1HW_.jpg   
1709  818145370475810820  https://pbs.twimg.com/media/C1qi26rW8AMaj9K.jpg   
2048  886983233522544640  https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
1987  872620804844003328  https://pbs.twimg.com/media/DBwr_hzXkAEnZBW.jpg   
685   683857920510050305  https://pbs.twimg.com/media/CX2NJmRWYAAxz_5.jpg   
1647  808733504066486276  https://pbs.twimg.com/media/Czky0v9VIAEXRkd.jpg   
116   668113020489474048  https://pbs.twimg.com/media/CUWdPsqWcAERQVv.jpg   
733   686749460672679938  https://pbs.twimg.com/media/CYfS75fWAAAllde.jpg   
438   674422304705744896  https://pbs.twimg.com/media/CVwHgblWcAACWOD.jpg   

      img_num                p1   p1_conf  p1_dog                  p2  \
2064        1           Samoyed  0.957979    True          Pomeranian   
460         1   giant_schnauzer  0.119693    True        Afghan_hound   
1709        1  golden_retriever  0.621931    True  Labrador_retriever   
2048        2         Chihuahua  0.793469    True         toy_terrier   
1987        1    cocker_spaniel  0.513191    True      Sussex_spaniel   
685         1          bluetick  0.174738    True   Shetland_sheepdog   
1647        1         seat_belt  0.779137   False          toy_poodle   
116         1          Pembroke  0.548896    True            Cardigan   
733         1      cheeseburger  0.643808   False              hotdog   
438         1  golden_retriever  0.964497    True  Labrador_retriever   

       p2_conf  p2_dog                   p3   p3_conf  p3_dog  
2064  0.013884    True                 chow  0.008167    True  
460   0.072763    True  miniature_schnauzer  0.063786    True  
1709  0.364997    True              redbone  0.003971    True  
2048  0.143528    True           can_opener  0.032253   False  
1987  0.159088    True      standard_poodle  0.149509    True  
685   0.126101    True               beagle  0.122887    True  
1647  0.036927    True     golden_retriever  0.016972    True  
116   0.191101    True               collie  0.059814    True  
733   0.201378   False                bagel  0.063880   False  
438   0.009006    True          tennis_ball  0.007139   False

In [131]:
image_predictions.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [132]:
image_predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [133]:
image_predictions[image_predictions.p1_dog == False]

tweet_id                                            jpg_url  \
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
33    666430724426358785    https://pbs.twimg.com/media/CT-jNYqW4AAPi2M.jpg   
43    666776908487630848    https://pbs.twimg.com/media/CUDeDoWUYAAD-EM.jpg   
45    666786068205871104    https://pbs.twimg.com/media/CUDmZIkWcAAIPPe.jpg   
50    666837028449972224    https://pbs.twimg.com/media/CUEUva1WsAA2jPb.jpg   
51    666983947667116034    https://pbs.twimg.com/media/CUGaXDhW4AY9JUH.jpg   
52    666996132027977728    https://pbs.twimg.com/media/CUGlb6iUwAITEbW.jpg   
53    667012601033924608    https://pbs.twimg.com/media/CUG0bC0U8AAw2su.jpg   
56    667065535570550784    https://pbs.twimg.com/media/CUHkkJpXIAA2w3n.jpg   
69    667188689915760640    https://pbs.twimg.com/media/CUJUk2iWUAAVtOv.jpg   
73    667369227918143488    https://pbs.twimg.com/media/CUL4xR9UkAEdlJ6.jpg   
77    667437278097252352    https://pbs.twimg.com/media/CUM2qWaWoAUZ06L.jpg   
78    667443425659232256    https://pbs.twimg.com/media/CUM8QZwW4AAVsBl.jpg   
87    667524857454854144    https://pbs.twimg.com/media/CUOGUfJW4AA_eni.jpg   
93    667549055577362432    https://pbs.twimg.com/media/CUOcVCwWsAERUKY.jpg   
94    667550882905632768    https://pbs.twimg.com/media/CUObvUJVEAAnYPF.jpg   
95    667550904950915073    https://pbs.twimg.com/media/CUOb_gUUkAACXdS.jpg   
96    667724302356258817    https://pbs.twimg.com/media/CUQ7tv3W4AA3KlI.jpg   
98    667766675769573376    https://pbs.twimg.com/media/CURiQMnUAAAPT2M.jpg   
100   667782464991965184    https://pbs.twimg.com/media/CURwm3cUkAARcO6.jpg   
103   667806454573760512    https://pbs.twimg.com/media/CUSGbXeVAAAgztZ.jpg   
106   667866724293877760    https://pbs.twimg.com/media/CUS9PlUWwAANeAD.jpg   
107   667873844930215936    https://pbs.twimg.com/media/CUTDtyGXIAARxus.jpg   
...                  ...                                                ...   
1900  851464819735769094    https://pbs.twimg.com/media/C9ECujZXsAAPCSM.jpg   
1902  851861385021730816    https://pbs.twimg.com/media/C8W6sY_W0AEmttW.jpg   
1904  852189679701164033    https://pbs.twimg.com/media/C9OV99SXsAEmj1U.jpg   
1905  852226086759018497  https://pbs.twimg.com/ext_tw_video_thumb/85222...   
1906  852311364735569921    https://pbs.twimg.com/media/C9QEqZ7XYAIR7fS.jpg   
1910  853299958564483072    https://pbs.twimg.com/media/C9eHyF7XgAAOxPM.jpg   
1931  859074603037188101    https://pbs.twimg.com/media/C-wLyufW0AA546I.jpg   
1932  859196978902773760  https://pbs.twimg.com/ext_tw_video_thumb/85919...   
1936  860184849394610176    https://pbs.twimg.com/media/C-_9jWWUwAAnwkd.jpg   
1937  860276583193509888    https://pbs.twimg.com/media/C_BQ_NlVwAAgYGD.jpg   
1940  860924035999428608    https://pbs.twimg.com/media/C_KVJjDXsAEUCWn.jpg   
1942  861288531465048066  https://pbs.twimg.com/ext_tw_video_thumb/86128...   
1944  861769973181624320    https://pbs.twimg.com/media/CzG425nWgAAnP7P.jpg   
1946  862457590147678208    https://pbs.twimg.com/media/C_gQmaTUMAAPYSS.jpg   
1953  863907417377173506    https://pbs.twimg.com/media/C_03NPeUQAAgrMl.jpg   
1956  864873206498414592    https://pbs.twimg.com/media/DAClmHkXcAA1kSv.jpg   
1970  868880397819494401    https://pbs.twimg.com/media/DA7iHL5U0AA1OQo.jpg   
1975  870063196459192321    https://pbs.twimg.com/media/DBMV3NnXUAAm0Pp.jpg   
1979  870804317367881728    https://pbs.twimg.com/media/DBW35ZsVoA

In [138]:
image_predictions[(image_predictions.p1_dog == False) & (image_predictions.p2_dog == False) & (image_predictions.p3_dog == False)]

tweet_id                                            jpg_url  \
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
45    666786068205871104    https://pbs.twimg.com/media/CUDmZIkWcAAIPPe.jpg   
50    666837028449972224    https://pbs.twimg.com/media/CUEUva1WsAA2jPb.jpg   
51    666983947667116034    https://pbs.twimg.com/media/CUGaXDhW4AY9JUH.jpg   
53    667012601033924608    https://pbs.twimg.com/media/CUG0bC0U8AAw2su.jpg   
56    667065535570550784    https://pbs.twimg.com/media/CUHkkJpXIAA2w3n.jpg   
69    667188689915760640    https://pbs.twimg.com/media/CUJUk2iWUAAVtOv.jpg   
73    667369227918143488    https://pbs.twimg.com/media/CUL4xR9UkAEdlJ6.jpg   
77    667437278097252352    https://pbs.twimg.com/media/CUM2qWaWoAUZ06L.jpg   
78    667443425659232256    https://pbs.twimg.com/media/CUM8QZwW4AAVsBl.jpg   
93    667549055577362432    https://pbs.twimg.com/media/CUOcVCwWsAERUKY.jpg   
94    667550882905632768    https://pbs.twimg.com/media/CUObvUJVEAAnYPF.jpg   
96    667724302356258817    https://pbs.twimg.com/media/CUQ7tv3W4AA3KlI.jpg   
98    667766675769573376    https://pbs.twimg.com/media/CURiQMnUAAAPT2M.jpg   
100   667782464991965184    https://pbs.twimg.com/media/CURwm3cUkAARcO6.jpg   
106   667866724293877760    https://pbs.twimg.com/media/CUS9PlUWwAANeAD.jpg   
107   667873844930215936    https://pbs.twimg.com/media/CUTDtyGXIAARxus.jpg   
112   667911425562669056    https://pbs.twimg.com/media/CUTl5m1WUAAabZG.jpg   
115   667937095915278337    https://pbs.twimg.com/media/CUT9PuQWwAABQv7.jpg   
117   668142349051129856    https://pbs.twimg.com/media/CUW37BzWsAAlJlN.jpg   
118   668154635664932864    https://pbs.twimg.com/media/CUXDGR2WcAAUQKz.jpg   
123   668226093875376128    https://pbs.twimg.com/media/CUYEFlQXAAUkPGm.jpg   
130   668291999406125056    https://pbs.twimg.com/media/CUZABzGW4AE5F0k.jpg   
132   668466899341221888    https://pbs.twimg.com/media/CUbfGbbWoAApZth.jpg   
140   668544745690562560    https://pbs.twimg.com/media/CUcl5jeWsAA6ufS.jpg   
...                  ...                                                ...   
1839  837482249356513284    https://pbs.twimg.com/media/C59VqMUXEAAzldG.jpg   
1844  838916489579200512    https://pbs.twimg.com/media/C6RkiQZUsAAM4R4.jpg   
1847  839290600511926273    https://pbs.twimg.com/media/C6XBt9XXEAEEW9U.jpg   
1851  840370681858686976    https://pbs.twimg.com/media/C6mYrK0UwAANhep.jpg   
1853  840696689258311684    https://pbs.twimg.com/media/C6rBLenU0AAr8MN.jpg   
1869  844580511645339650    https://pbs.twimg.com/media/C7iNfq1W0AAcbsR.jpg   
1886  847962785489326080    https://pbs.twimg.com/media/C8SRpHNUIAARB3j.jpg   
1887  847971574464610304    https://pbs.twimg.com/media/C8SZH1EWAAAIRRF.jpg   
1891  849051919805034497    https://pbs.twimg.com/media/C8hwNxbXYAAwyVG.jpg   
1892  849336543269576704    https://pbs.twimg.com/media/C8lzFC4XcAAQxB4.jpg   
1900  851464819735769094    https://pbs.twimg.com/media/C9ECujZXsAAPCSM.jpg   
1902  851861385021730816    https://pbs.twimg.com/media/C8W6sY_W0AEmttW.jpg   
1905  852226086759018497  https://pbs.twimg.com/ext_tw_video_thumb/85222...   
1906  852311364735569921    https://pbs.twimg.com/media/C9QEqZ7XYAIR7fS.jpg   
1910  853299958564483072    https://pbs.twimg.com/media/C9eHyF7XgAAOxPM.jpg   
1931  859074603037188101    https://pbs.twimg.com/media/C-wLyufW0AA546I.jpg   
1936  860184849394610176    https://pbs.twimg.com/media/C-_9jWWUwAAnwkd.jpg   
1937  860276583193509888    https://pbs.twimg.com/media/C_BQ_NlVwAAgYGD.jpg   
1940  860924035999428608    https://pbs.twimg.com/media/C_KVJjDXsA

#### Analyze the `detailed_tweets` table

In [139]:
detailed_tweets.sample(10)

contributors  coordinates          created_at  \
1037           NaN          NaN 2016-06-13 01:06:33   
2018           NaN          NaN 2015-12-03 02:45:32   
2099           NaN          NaN 2015-11-28 19:04:19   
2109           NaN          NaN 2015-11-25 01:20:08   
1616           NaN          NaN 2016-01-08 19:45:39   
2147           NaN          NaN 2015-11-24 02:29:49   
1550           NaN          NaN 2016-01-12 02:06:41   
1192           NaN          NaN 2016-03-14 02:39:42   
2048           NaN          NaN 2015-12-01 19:10:13   
1908           NaN          NaN 2015-12-07 03:45:53   

                                               entities  \
1037  {'hashtags': [], 'symbols': [], 'user_mentions...   
2018  {'hashtags': [], 'symbols': [], 'user_mentions...   
2099  {'hashtags': [], 'symbols': [], 'user_mentions...   
2109  {'hashtags': [], 'symbols': [], 'user_mentions...   
1616  {'hashtags': [], 'symbols': [], 'user_mentions...   
2147  {'hashtags': [], 'symbols': [], 'user_mentions...   
1550  {'hashtags': [], 'symbols': [], 'user_mentions...   
1192  {'hashtags': [], 'symbols': [], 'user_mentions...   
2048  {'hashtags': [], 'symbols': [], 'user_mentions...   
1908  {'hashtags': [], 'symbols': [], 'user_mentions...   

                                      extended_entities  favorite_count  \
1037  {'media': [{'id': 742161185223630852, 'id_str'...            4468   
2018  {'media': [{'id': 672245248555409408, 'id_str'...             680   
2099  {'media': [{'id': 670679606182055936, 'id_str'...             745   
2109  {'media': [{'id': 669324644248219648, 'id_str'...             494   
1616  {'media': [{'id': 685547930707750912, 'id_str'...           33890   
2147  {'media': [{'id': 668979801907175424, 'id_str'...             793   
1550  {'media': [{'id': 686730987418550272, 'id_str'...            4262   
1192  {'media': [{'id': 709207338792525824, 'id_str'...           12890   
2048  {'media': [{'id': 671768277677441024, 'id_str'...            1184   
1908  {'media': [{'id': 673709986195890176, 'id_str'...             849   

      favorited  geo                  id              id_str  ...  \
1037      False  NaN  742161199639494656  742161199639494656  ...   
2018      False  NaN  672245253877968896  672245253877968896  ...   
2099      False  NaN  670679630144274432  670679630144274432  ...   
2109      False  NaN  669324657376567296  669324657376567296  ...   
1616      False  NaN  685547936038666240  685547936038666240  ...   
2147      False  NaN  668979806671884288  668979806671884288  ...   
1550      False  NaN  686730991906516992  686730991906516992  ...   
1192      False  NaN  709207347839836162  709207347839836160  ...   
2048      False  NaN  671768281401958400  671768281401958400  ...   
1908      False  NaN  673709992831262724  673709992831262720  ...   

     quoted_status  quoted_status_id  quoted_status_id_str  retweet_count  \
1037           NaN               NaN                   NaN           1429   
2018           NaN               NaN                   NaN            159   
2099           NaN               NaN                   NaN            278   
2109           NaN               NaN                   NaN            207   
1616           NaN               NaN                   NaN          16185   
2147           NaN               NaN                   NaN            348   
1550           NaN               NaN                   NaN           1235   
1192           NaN               NaN                   NaN           5980   
2048           NaN               NaN                   NaN            507   
1908           NaN               NaN                   NaN            275   

      retweeted  retweeted_status  \
1037      False               NaN   
2018      False               NaN   
2099      False               NaN   
2109      False               NaN   
1616      False               NaN   
2147      False               NaN   
1550      False               NaN   
1192      False     

In [140]:
detailed_tweets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2333 entries, 0 to 2332
Data columns (total 29 columns):
contributors                 0 non-null float64
coordinates                  0 non-null float64
created_at                   2333 non-null datetime64[ns]
entities                     2333 non-null object
extended_entities            1818 non-null object
favorite_count               2333 non-null int64
favorited                    2333 non-null bool
geo                          0 non-null float64
id                           2333 non-null int64
id_str                       2333 non-null int64
in_reply_to_screen_name      77 non-null object
in_reply_to_status_id        77 non-null float64
in_reply_to_status_id_str    77 non-null float64
in_reply_to_user_id          77 non-null float64
in_reply_to_user_id_str      77 non-null float64
is_quote_status              2333 non-null bool
lang                         2333 non-null object
place                        1 non-null object
possibl

In [141]:
detailed_tweets.describe()

contributors  coordinates  favorite_count  geo            id  \
count           0.0          0.0     2333.000000  0.0  2.333000e+03   
mean            NaN          NaN     7781.912130  NaN  7.419023e+17   
std             NaN          NaN    12072.355158  NaN  6.818084e+16   
min             NaN          NaN        0.000000  NaN  6.660209e+17   
25%             NaN          NaN     1355.000000  NaN  6.782786e+17   
50%             NaN          NaN     3390.000000  NaN  7.184547e+17   
75%             NaN          NaN     9545.000000  NaN  7.986440e+17   
max             NaN          NaN   161207.000000  NaN  8.924206e+17   

             id_str  in_reply_to_status_id  in_reply_to_status_id_str  \
count  2.333000e+03           7.700000e+01               7.700000e+01   
mean   7.419023e+17           7.440692e+17               7.440692e+17   
std    6.818084e+16           7.524295e+16               7.524295e+16   
min    6.660209e+17           6.658147e+17               6.658147e+17   
25%    6.782786e+17           6.757073e+17               6.757073e+17   
50%    7.184547e+17           7.032559e+17               7.032559e+17   
75%    7.986440e+17           8.233264e+17               8.233264e+17   
max    8.924206e+17           8.862664e+17               8.862664e+17   

       in_reply_to_user_id  in_reply_to_user_id_str  possibly_sensitive  \
count         7.700000e+01             7.700000e+01              2199.0   
mean          2.040329e+16             2.040329e+16                 0.0   
std           1.260797e+17             1.260797e+17                 0.0   
min           1.185634e+07             1.185634e+07                 0.0   
25%           3.589728e+08             3.589728e+08                 0.0   
50%           4.196984e+09             4.196984e+09                 0.0   
75%           4.196984e+09             4.196984e+09                 0.0   
max           8.405479e+17             8.405479e+17                 0.0   

       quoted_status_id  quoted_status_id_str  retweet_count  
count      2.600000e+01          2.600000e+01    2333.000000  
mean       8.113972e+17          8.113972e+17    2814.629233  
std        6.295843e+16          6.295843e+16    4760.842662  
min        6.721083e+17          6.721083e+17       1.000000  
25%        7.761338e+17          7.761338e+17     565.000000  
50%        8.281173e+17          8.281173e+17    1316.000000  
75%        8.637581e+17          8.637581e+17    3281.000000  
max        8.860534e+17          8.860534e+17   80810.000000

In [142]:
detailed_tweets[detailed_tweets.favorited == True]

Empty DataFrame
Columns: [contributors, coordinates, created_at, entities, extended_entities, favorite_count, favorited, geo, id, id_str, in_reply_to_screen_name, in_reply_to_status_id, in_reply_to_status_id_str, in_reply_to_user_id, in_reply_to_user_id_str, is_quote_status, lang, place, possibly_sensitive, quoted_status, quoted_status_id, quoted_status_id_str, retweet_count, retweeted, retweeted_status, source, text, truncated, user]
Index: []

[0 rows x 29 columns]

In [143]:
detailed_tweets[detailed_tweets.retweeted == True]

Empty DataFrame
Columns: [contributors, coordinates, created_at, entities, extended_entities, favorite_count, favorited, geo, id, id_str, in_reply_to_screen_name, in_reply_to_status_id, in_reply_to_status_id_str, in_reply_to_user_id, in_reply_to_user_id_str, is_quote_status, lang, place, possibly_sensitive, quoted_status, quoted_status_id, quoted_status_id_str, retweet_count, retweeted, retweeted_status, source, text, truncated, user]
Index: []

[0 rows x 29 columns]

In [144]:
detailed_tweets[detailed_tweets.is_quote_status == True]

contributors  coordinates          created_at  \
6              NaN          NaN 2017-06-24 13:24:20   
12             NaN          NaN 2017-07-13 15:19:09   
54             NaN          NaN 2017-07-15 02:45:48   
68             NaN          NaN 2017-06-18 20:30:39   
75             NaN          NaN 2017-06-14 21:06:43   
93             NaN          NaN 2017-07-10 03:08:17   
114            NaN          NaN 2017-05-22 18:21:28   
138            NaN          NaN 2017-04-22 16:18:34   
142            NaN          NaN 2017-06-03 20:33:19   
157            NaN          NaN 2017-04-22 18:55:51   
218            NaN          NaN 2017-03-07 03:22:35   
224            NaN          NaN 2017-03-13 16:08:50   
260            NaN          NaN 2017-03-26 23:20:02   
271            NaN          NaN 2017-03-21 00:22:10   
275            NaN          NaN 2017-03-27 23:35:28   
333            NaN          NaN 2017-02-17 20:05:43   
345            NaN          NaN 2017-02-21 17:18:39   
387            NaN          NaN 2017-01-24 22:44:42   
400            NaN          NaN 2017-01-08 01:40:55   
492            NaN          NaN 2016-12-02 00:02:45   
506            NaN          NaN 2016-11-22 00:17:10   
574            NaN          NaN 2016-12-07 19:09:37   
629            NaN          NaN 2016-10-21 03:56:25   
671            NaN          NaN 2016-11-08 23:01:49   
798            NaN          NaN 2016-07-30 17:51:13   
838            NaN          NaN 2016-09-03 03:13:29   
1055           NaN          NaN 2016-06-18 17:41:06   
1081           NaN          NaN 2016-04-27 22:57:10   
1185           NaN          NaN 2016-03-22 16:06:19   
1270           NaN          NaN 2016-04-03 20:53:33   
1307           NaN          NaN 2016-03-05 17:26:40   

                                               entities extended_entities  \
6     {'hashtags': [], 'symbols': [], 'user_mentions...               NaN   
12    {'hashtags': [], 'symbols': [], 'user_mentions...               NaN   
54    {'hashtags': [{'text': 'BATP', 'indices': [21,...               NaN   
68    {'hashtags': [], 'symbols': [], 'user_mentions...               NaN   
75    {'hashtags': [], 'symbols': [], 'user_mentions...               NaN   
93    {'hashtags': [], 'symbols': [], 'user_mentions...               NaN   
114   {'hashtags': [], 'symbols': [], 'user_mentions...               NaN   
138   {'hashtags': [], 'symbols': [], 'user_mentions...               NaN   
142   {'hashtags': [], 'symbols': [], 'user_mentions...               NaN   
157   {'hashtags': [], 'symbols': [], 'user_mentions...               NaN   
218   {'hashtags': [], 'symbols': [], 'user_mentions...               NaN   
224   {'hashtags': [], 'symbols': [], 'user_mentions...               NaN   
260   {'hashtags': [], 'symbols': [], 'user_mentions...               NaN   
271   {'hashtags': [], 'symbols': [], 'user_mentions...               NaN   
275   {'hashtags': [], 'symbols': [], 'user_mentions...               NaN   
333   {'hashtags': [], 'symbols': [], 'user_mentions...               NaN   
345   {'hashtags': [], 'symbols': [], 'user_mentions...               NaN   
387   {'hashtags': [], 'symbols': [], 'user_mentions...               NaN   
400   {'hashtags': [], 'symbols': [], 'user_mentions...               NaN   
492   {'hashtags': [], 'symbols': [], 'user_mentions...               NaN   
506   {'hashtags': [], 'symbols': [], 'user_mentions...               NaN   
574   {'hashtags': [], 'symbols': [], 'user_mentions...               NaN   
629   {'hashtags': [], 'symbols': [], 'user_mentions...               NaN   
671   {'hashtags': [{'text': 'ImWithThor', 'indices'...               NaN   
798   {'hashtags': [], 'symbols': [], 'user_mentions...               NaN   
838   {'hashtags': [], 'symbols': [], 'user_mentions...               NaN   
1055  {'hashtags': [], 'symbols': [], 'user_mentions...               NaN   
1081  {'hashtags': [], 'symbols': [], 'user_mentions...               NaN   
1185  {'hash

In [147]:
detailed_tweets[detailed_tweets.place.notnull()]

contributors  coordinates          created_at  \
837           NaN          NaN 2016-08-10 01:23:03   

                                              entities  \
837  {'hashtags': [], 'symbols': [], 'user_mentions...   

                                     extended_entities  favorite_count  \
837  {'media': [{'id': 763183833575481344, 'id_str'...            5599   

     favorited  geo                  id              id_str  ...  \
837      False  NaN  763183847194451968  763183847194451968  ...   

    quoted_status  quoted_status_id  quoted_status_id_str  retweet_count  \
837           NaN               NaN                   NaN           1546   

     retweeted  retweeted_status  \
837      False               NaN   

                                                source  \
837  <a href="http://twitter.com/download/iphone" r...   

                                                  text  truncated  \
837  This is Clark. He collects teddy bears. It's a...      False   

                                                  user  
837  {'id': 4196983835, 'id_str': '4196983835', 'na...  

[1 rows x 29 columns]

In [152]:
detailed_tweets.iloc[837]

contributors                                                               NaN
coordinates                                                                NaN
created_at                                                 2016-08-10 01:23:03
entities                     {'hashtags': [], 'symbols': [], 'user_mentions...
extended_entities            {'media': [{'id': 763183833575481344, 'id_str'...
favorite_count                                                            5599
favorited                                                                False
geo                                                                        NaN
id                                                          763183847194451968
id_str                                                      763183847194451968
in_reply_to_screen_name                                                   None
in_reply_to_status_id                                                      NaN
in_reply_to_status_id_str                           

In [159]:
detailed_tweets[detailed_tweets.in_reply_to_screen_name.notnull()]

contributors  coordinates          created_at  \
7              NaN          NaN 2017-06-27 12:14:36   
21             NaN          NaN 2017-07-15 16:51:35   
38             NaN          NaN 2017-07-02 21:58:53   
103            NaN          NaN 2017-04-22 19:05:32   
115            NaN          NaN 2017-05-13 16:15:35   
136            NaN          NaN 2017-05-12 17:12:53   
151            NaN          NaN 2017-04-24 15:13:52   
169            NaN          NaN 2017-04-22 19:15:32   
183            NaN          NaN 2017-04-26 12:48:51   
186            NaN          NaN 2017-06-02 19:38:25   
206            NaN          NaN 2017-03-31 01:11:22   
212            NaN          NaN 2017-03-23 18:29:57   
230            NaN          NaN 2017-03-11 22:59:09   
238            NaN          NaN 2017-04-01 16:41:12   
248            NaN          NaN 2017-03-04 17:56:49   
253            NaN          NaN 2017-03-04 22:12:52   
273            NaN          NaN 2017-04-07 13:04:55   
288            NaN          NaN 2017-02-16 04:45:50   
307            NaN          NaN 2017-02-01 01:11:25   
349            NaN          NaN 2017-02-24 21:54:03   
376            NaN          NaN 2017-02-15 18:03:45   
388            NaN          NaN 2016-12-25 21:12:41   
391            NaN          NaN 2017-01-23 00:56:15   
429            NaN          NaN 2017-01-17 00:33:26   
493            NaN          NaN 2016-11-25 21:37:47   
532            NaN          NaN 2016-11-24 18:28:13   
534            NaN          NaN 2016-11-22 00:32:18   
542            NaN          NaN 2016-12-21 19:01:02   
649            NaN          NaN 2016-11-11 19:55:50   
700            NaN          NaN 2016-10-12 03:50:17   
...            ...          ...                 ...   
1457           NaN          NaN 2016-02-06 00:35:13   
1469           NaN          NaN 2016-01-27 19:05:49   
1478           NaN          NaN 2016-01-10 04:04:10   
1515           NaN          NaN 2016-01-27 00:31:15   
1517           NaN          NaN 2016-01-22 18:49:36   
1579           NaN          NaN 2015-12-28 05:07:27   
1582           NaN          NaN 2016-01-01 06:22:03   
1591           NaN          NaN 2016-01-09 04:34:45   
1652           NaN          NaN 2016-01-05 04:11:44   
1654           NaN          NaN 2016-01-07 05:28:35   
1676           NaN          NaN 2016-01-06 00:54:18   
1693           NaN          NaN 2015-12-19 01:25:31   
1785           NaN          NaN 2015-12-11 16:20:15   
1791           NaN          NaN 2015-12-10 17:11:09   
1792           NaN          NaN 2015-12-13 02:51:51   
1814           NaN          NaN 2015-12-13 01:25:37   
1831           NaN          NaN 2015-12-12 16:02:36   
1838           NaN          NaN 2015-12-15 02:32:17   
1860           NaN          NaN 2015-12-10 00:54:28   
1874           NaN          NaN 2015-12-10 03:30:58   
1875           NaN          NaN 2015-12-10 00:08:50   
1958           NaN          NaN 2015-12-09 15:09:55   
1973           NaN          NaN 2015-12-07 04:11:02   
1974           NaN          NaN 2015-12-08 20:53:11   
2011           NaN          NaN 2015-12-01 16:37:44   
2028           NaN          NaN 2015-12-01 04:44:10   
2082           NaN          NaN 2015-11-25 03:14:30   
2102           NaN          NaN 2015-11-26 01:11:28   
2155           NaN          NaN 2015-11-24 01:42:25   
2270           NaN          NaN 2015-11-18 20:02:51   

                                               entities  \
7     {'hashtags': [], 'symbols': [], 'user_mentions...   
21    {'hashtags': [], 'symbols': [], 'user_mentions...   
38    {'hashtags': [], 'symbols': [], 'user_mentions...   
103   {'hashtags': [], 'symbols': [], 'user_mentions...   
115   {'hashtags': [], 'symbols': [], 'user_mentions...   
136   {'hashtags': [], 'symbols': [], 'user_mentions...   
151   {'hashtags': [], 'symbols': [], 'user_mentions...   
169   {'hashtags': [], 'symbols': [], 'user_mentions...   
183   {'hashtags': [], 'symbols': [], 'user_mentions...   
186  

In [160]:
detailed_tweets.iloc[21]

contributors                                                               NaN
coordinates                                                                NaN
created_at                                                 2017-07-15 16:51:35
entities                     {'hashtags': [], 'symbols': [], 'user_mentions...
extended_entities                                                          NaN
favorite_count                                                             116
favorited                                                                False
geo                                                                        NaN
id                                                          886267009285017600
id_str                                                      886267009285017600
in_reply_to_screen_name                                            NonWhiteHat
in_reply_to_status_id                                              8.86266e+17
in_reply_to_status_id_str                           

### Quality
#### `twitter_archive` table
- 78 non-original tweets which contains in_reply_to_status_id, in_reply_to_user_id
- in_reply_to_status_id is a float, could be an int
- in_reply_to_user_id is a float, could be a string
- timestamp is an object, should be a datetime
- retweeted_status_id is a float, could be a string
- retweeted_status_user_id is a float, could be a string
- retweeted_status_timestamp is an object, should be a datetime
- 181 retweets present in table - contains retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp
- missing records in expanded_urls
- some names have the value None
- some rows have no entries in either of doggo, floofer, pupper, puppo
- ~~some rating_numerators have value 0~~
- ~~some rating_numerators have very large values~~
- ~~some rating_denominators have value 0~~
- ~~some rating_denominators have very large values~~
- some of the animals (with low scores) are not dogs (need to decide how to process these)
- some of the animals (with very high scores) are in fact dogs and valid tweets

#### `image_predictions` table
- record 1647 is that of a dog, but p1 finds it a seat belt
- record 1906 is clearly a dog, but p1/p2/p3 finds it otherwise
- record 1953 is a dog in the bushes, but identification is otherwise

#### `detailed_tweets` table
- contributors has empty values
- coordinates has empty values
- geo has empty values
- place has empty values
- in_reply_to_status_id is a float, could be a string
- quoted_status_id is a float, could be a string

### Tidiness
#### `twitter_archive` table
- Dog 'state' is split into 4 columns - doggo, floofer, pupper, puppo
- `image_predictions` could be combined into `twitter_archive` table
- `detailed_tweets` could be combined into `twitter_archive` table
- contributors is unnecessary column
- coordinates is unnecessary column
- geo is unnecessary column

